src/ir/context.rs

```diff
@@ -1794,6 +1811,48 @@ impl<'ctx> BindgenContext<'ctx> {
         // derive default or not.
         !self.cannot_derive_default.as_ref().unwrap().contains(&id)
     }
+
+    /// Compute whether we can derive debug.
+    fn compute_cannot_derive_copy(&mut self) {
+        assert!(self.cannot_derive_copy.is_none());
+        let (cant_derive_copy, cant_derive_copy_in_array) =
+            analyze::<CannotDeriveCopy>(self);
+        self.cannot_derive_copy = Some(cant_derive_copy);
+        self.cannot_derive_copy_in_array = Some(cant_derive_copy_in_array);
+    }
+
+    /// Look up whether the item with `id` can
+    /// derive debug or not.
+    pub fn lookup_item_id_can_derive_copy(&self, id: ItemId) -> bool {
+        assert!(self.in_codegen_phase(),
+                "We only compute can_derive_debug when we enter codegen");
+
+        let contains_array = id.has_array(self);
+        // Look up the computed value for whether the item with `id` can
+        // derive `Copy` or not.
+        trace!("looking up id: {:?}, contains_array: {:?}", id, contains_array);
+        if contains_array {
+            !self.cannot_derive_copy_in_array.as_ref().unwrap().contains(&id)
```

Comment:
Is there ever a case where the type can have an array containing a named type parameter and it is still true that it can derive copy? As far as I know, the answer is "no", which means that this check is already unnecessary. That is, if `id.has_array(ctx)` then we can return `false` immediately. Which also means that we don't need the `cannot_derive_copy_in_array` set, because its redundant with the `HasArray` analysis's results.

So, unless I'm overlooking something, I think we should take these steps:

* Rename `HasArrayAnalysis` to `HasTypeParameterInArray` so that it is more clear what it is computing. It isn't computing if the type has any old array in it, it is computing whether it contains an array of type parameters. We can drop the `Analysis` because it should be obvious from the context: its in the `ir::analysis` module amd it implements the `MonotonFramework` trait.

* Only compute the `cannot_derive_copy` set in the `CannotDeriveCopy` analysis. Get rid of the `cannot_derive_copy_in_array` set.

* In this method:

```rust
assert!(... in codegen phase ...);
!id.has_type_param_in_array(self) &&
    !self.cannot_derive_copy.as_ref().unwrap().contains(&id)
```

Sorry for not catching this earlier! I think this will be a big improvement, though, so it is worth it.